### ISSUE TITLE
List of known differences with old Tensorflow implementation

### ISSUE THREAD

**Issue by melisande-c** [CAREamics/careamics#504]
There are some known differences in implementation between the old `n2v` package and `CAREamics`. We should thoroughly benchmark any differences in performance.

### Upsampling method

The `n2v` package used nearest-neighbour upsampling whereas CAREamics uses bilinear or trilinear. 

In some limited tests that I ran, I noticed N2V2 seems to converge more quickly with bilinear upsampling but nearest neighbour seems so enforce a smoother output function.

### Batch normalisation parameters

The default parameters for batch normalisation are different between Tensorflow and PyTorch. To make PyTorch `BatchNorm` equivalent to Tensorflow's we need to set `momentum=0.01` and `eps=0.001`, (PyTorch defines momentum as 1 - Tensorflow's momentum).

### Weight initialisation 

PyTorch uses Kaiming Uniform weight initialisation and Tensorflow uses Glorot Uniform weight initialisation, by default.

### Max blur pool layer

The outputs from `CAREamics` and the `n2v` package are very slightly different.

### Decoder conv block intermediate feature multiplier

In the `n2v` package when the parameter `skip_skipone=False` the decoder conv block intermediate feature multiplier is equal to 1, but when `skip_skipone=False` it is equal to 2.

Decoder summary with `skip_skipone=False` (depth=2):
```
 up_sampling2d (UpSampling2  (None, None, None, 64)       0         ['activation_5[0][0]']        
 D)                                                                                               
                                                                                                  
 concatenate (Concatenate)   (None, None, None, 128)      0         ['up_sampling2d[0][0]',       
                                                                     'activation_3[0][0]']        
                                                                                                  
 channel_0up_level_1_no_0 (  (None, None, None, 32)       36896     ['concatenate[0][0]']         
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_6 (Bat  (None, None, None, 32)       128       ['channel_0up_level_1_no_0[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_6 (Activation)   (None, None, None, 32)       0         ['batch_normalization_6[0][0]'
                                                                    ]                             
                                                                                                  
 channel_0up_level_1_no_2 (  (None, None, None, 32)       9248      ['activation_6[0][0]']        
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_7 (Bat  (None, None, None, 32)       128       ['channel_0up_level_1_no_2[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_7 (Activation)   (None, None, None, 32)       0         ['batch_normalization_7[0][0]'
                                                                    ]                             
                                                                                                  
 up_sampling2d_1 (UpSamplin  (None, None, None, 32)       0         ['activation_7[0][0]']        
 g2D)                                                                                             
                                                                                                  
 concatenate_1 (Concatenate  (None, None, None, 64)       0         ['up_sampling2d_1[0][0]',     
 )                                                                   'activation_1[0][0]']        
                                                                                                  
 channel_0up_level_0_no_0 (  (None, None, None, 32)       18464     ['concatenate_1[0][0]']       
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_8 (Bat  (None, None, None, 32)       128       ['channel_0up_level_0_no_0[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_8 (Activation)   (None, None, None, 32)       0         ['batch_normalization_8[0][0]'
                                                                    ]                             
                                                                                                  
 channel_0up_level_0_no_2 (  (None, None, None, 32)       9248      ['activation_8[0][0]']        
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_9 (Bat  (None, None, None, 32)       128       ['channel_0up_level_0_no_2[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_9 (Activation)   (None, None, None, 32)       0         ['batch_normalization_9[0][0]'
                                                                    ]                             
                                                                                                  
 conv2d (Conv2D)             (None, None, None, 1)        33        ['activation_9[0][0]']        
                                                                                                  
 activation_10 (Activation)  (None, None, None, 1)        0         ['conv2d[0][0]']  
```

Decoder summary with `skip_skipone=True` (depth=2):
```
 up_sampling2d (UpSampling2  (None, None, None, 64)       0         ['activation_5[0][0]']        
 D)                                                                                               
                                                                                                  
 concatenate (Concatenate)   (None, None, None, 128)      0         ['up_sampling2d[0][0]',       
                                                                     'activation_3[0][0]']        
                                                                                                  
 channel_0up_level_1_no_0 (  (None, None, None, 64)       73792     ['concatenate[0][0]']         
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_6 (Bat  (None, None, None, 64)       256       ['channel_0up_level_1_no_0[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_6 (Activation)   (None, None, None, 64)       0         ['batch_normalization_6[0][0]'
                                                                    ]                             
                                                                                                  
 channel_0up_level_1_no_2 (  (None, None, None, 32)       18464     ['activation_6[0][0]']        
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_7 (Bat  (None, None, None, 32)       128       ['channel_0up_level_1_no_2[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_7 (Activation)   (None, None, None, 32)       0         ['batch_normalization_7[0][0]'
                                                                    ]                             
                                                                                                  
 up_sampling2d_1 (UpSamplin  (None, None, None, 32)       0         ['activation_7[0][0]']        
 g2D)                                                                                             
                                                                                                  
 channel_0up_level_0_no_0 (  (None, None, None, 32)       9248      ['up_sampling2d_1[0][0]']     
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_8 (Bat  (None, None, None, 32)       128       ['channel_0up_level_0_no_0[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_8 (Activation)   (None, None, None, 32)       0         ['batch_normalization_8[0][0]'
                                                                    ]                             
                                                                                                  
 channel_0up_level_0_no_2 (  (None, None, None, 32)       9248      ['activation_8[0][0]']        
 Conv2D)                                                                                          
                                                                                                  
 batch_normalization_9 (Bat  (None, None, None, 32)       128       ['channel_0up_level_0_no_2[0][
 chNormalization)                                                   0]']                          
                                                                                                  
 activation_9 (Activation)   (None, None, None, 32)       0         ['batch_normalization_9[0][0]'
                                                                    ]                             
                                                                                                  
 conv2d (Conv2D)             (None, None, None, 1)        33        ['activation_9[0][0]']        
                                                                                                  
 activation_10 (Activation)  (None, None, None, 1)        0         ['conv2d[0][0]']   
```
